Develop gps

silviu-toderita opened the pull request:
Closes #32 with the following changes:

- i2c speed increased from 100 KHz to 400KHz. This has hugely sped up the polling of data, and will require each Sparkfun GNSS we use to have its pull-up resistors on the i2c bus disconnected by cutting a jumper. 
- Now polls data in the background at a rate set in the constructor. Each poll takes about 6ms, so currently set to 4 Hz for a total of 24ms blocking per second. This isn't so much a block as it is the time required to receive all data over i2c. A lot of data is sent out each time there is an update, and there is no way to further limit the data received. 
- Getters return locally-cached data
- Removed NMEA messages completely, meaning RMC sentences are no longer available (this decreased the amount of data we receive over i2c bus)
- Added Latitude / Longitude
- Added Heading
- Horizontal Speed now calculated by GPS module and pulled directly
- Horizontal Acceleration derived in this library
- Added Horizontal and Vertical Accuracy
- Added Altitude
- Vertical Speed and Vertical Acceleration derived in this library

silviu-toderita commented on lib/Telemetry/src/SensorGps.cpp:
```diff
@@ -1,81 +1,115 @@
-
-SensorGps::SensorGps(uint16_t updateInterval) {
-    _updateInterval = updateInterval;
     _gps = new SFE_UBLOX_GNSS();
```
I learned about this in a Java class, but never done it in C++. I believe I implemented it correctly, but maybe you can double check? @cosparks @valentino-jaber might also know better than me.

lcfyi commented on lib/Telemetry/src/SensorGps.cpp:
```diff
@@ -1,81 +1,119 @@
+SensorGps& SensorGps::operator = (const SensorGps &other){
+    if(this == &other) return *this;
+    _updateFrequency = other._updateFrequency;
+    _gps = other._gps;
```
I think this line may be problematic if we have a situation like so:

```cpp
SensorGps a = new SensorGps(...);
SensorGps b = new SensorGps(...);
a = b;
delete b;
// a now holds a reference to an internal `_gps` that has been freed
```

It may be more correct to just pass in the `SFE_UBLOX_GNSS` dependency that's constructed externally; there isn't necessarily a problem with passing in physical resources across multiple instances of a sensor.

lcfyi reviewed (changes requested):
I added a comment on the copy assignment operator; on further thought, it may be more correct to pass in the GPS dependency just so we don't implicitly reconstruct an interface to a physical resource or accidentally free something if we need multiple sensors for whatever reason.

silviu-toderita commented:
> Looks good to me, I'll leave it up to you on whether you want to include the other constructors etc.
> 
> I'm curious, have you done a live test with the GPS by publishing data to the Particle Cloud? I'd be interested in seeing some real data in our database so I can write a SQL query to display the path on a map.

Not yet, because I'm waiting for @cosparks to develop a new JSON schema as part of the issue he's working on. As soon as that's done, we'll upload some test data.

silviu-toderita commented:
> I added a comment on the copy assignment operator; on further thought, it may be more correct to pass in the GPS dependency just so we don't implicitly reconstruct an interface to a physical resource or accidentally free something if we need multiple sensors for whatever reason.

I'm a little confused, do you mean that this class should have an object of SFE_UBLOX_GNSS class passed in as an argument to the constructor? And then this would eliminate the need for the destructor and other functions I added?

lcfyi commented:
> Not yet, because I'm waiting for @cosparks to develop a new JSON schema as part of the issue he's working on. As soon as that's done, we'll upload some test data.

If you get a chance, could you publish some data with the old schema? The cloud infra currently expects the old schema so I wanted to see what sort of mapping panels we could add to the Grafana instance.

> I'm a little confused, do you mean that this class should have an object of SFE_UBLOX_GNSS class passed in as an argument to the constructor? And then this would eliminate the need for the destructor and other functions I added?

Yes. Since you're passing in a pointer to the dependency, you wouldn't have to worry about the rule of three. It's a [useful pattern](https://en.wikipedia.org/wiki/Dependency_injection) that complements composition when working with rather complex objects.

silviu-toderita commented:
> > Not yet, because I'm waiting for @cosparks to develop a new JSON schema as part of the issue he's working on. As soon as that's done, we'll upload some test data.
> 
> If you get a chance, could you publish some data with the old schema? The cloud infra currently expects the old schema so I wanted to see what sort of mapping panels we could add to the Grafana instance.
> 
> > I'm a little confused, do you mean that this class should have an object of SFE_UBLOX_GNSS class passed in as an argument to the constructor? And then this would eliminate the need for the destructor and other functions I added?
> 
> Yes. Since you're passing in a pointer to the dependency, you wouldn't have to worry about the rule of three. It's a [useful pattern](https://en.wikipedia.org/wiki/Dependency_injection) that complements composition when working with rather complex objects.

Done! As for the GPS data, I'll flash an old commit and let you know when I've uploaded some data.
